### PR TITLE
[WIP] Fix docked collection expressions in let-bindings

### LINKED ISSUE #967: Bug: break-separators=after-and-docked doesn't affect type definitions
```
type t =
  { hash: hash_type;
    file_mode: mode option;
    funs: id list;
    classes: id list;
    typedefs: id list;
    consts: id list;
    comments: (Pos.t * comment) list option }

let empty_t =
  { hash= None;
    file_mode= None;
    funs= [];
    classes= [];
    typedefs= [];
    consts= [];
    comments= Some [] }
```

Produces:

```
$ ocamlformat --break-separators=after-and-docked demo1.ml
type t = {
  hash: hash_type;
  file_mode: mode option;
  funs: id list;
  classes: id list;
  typedefs: id list;
  consts: id list;
  comments: (Pos.t * comment) list option;
}

let empty_t =
  {
    hash= None;
    file_mode= None;
    funs= [];
    classes= [];
    typedefs= [];
    consts= [];
    comments= Some [];
  }
```

This is inconsistent with the other values of `break-separators`, which do affect type definitions.

### LINKED ISSUE #1009: Bug: break-separators=after-and-docked doesn't apply to tuples
First thanks for the continued improvement of `ocamlformat`.

**Describe the bug**
`break-separators=after-and-docked` should apply to all bracketed expressions/patterns (i.e. records/lists/arrays/tuples). https://github.com/ocaml-ppx/ocamlformat/pull/931 fixed it for lists/arrays, however tuples are still not affected.

**How to Reproduce**

```ocaml
[@@@ocamlformat "break-separators=after-and-docked"]

let x =
  ( "foofoofoofoofoofoofoofoofoofoo",
    "barbarbarbarbarbarbarbarbarbar",
    "foobarfoobarfoobarfoobarfoobar" )
```

I would expect something like

```ocaml
[@@@ocamlformat "break-separators=after-and-docked"]

let x =
  (
    "foofoofoofoofoofoofoofoofoofoo",
    "barbarbarbarbarbarbarbarbarbar",
    "foobarfoobarfoobarfoobarfoobar"
  )
```

or

```ocaml
[@@@ocamlformat "break-separators=after-and-docked"]

let x = (
  "foofoofoofoofoofoofoofoofoofoo",
  "barbarbarbarbarbarbarbarbarbar",
  "foobarfoobarfoobarfoobarfoobar"
)
```

### PR DESCRIPTION
**Pull request by gpetiot**
Replaces #1008 
Fix #967 also fix #1009

### CODE REVIEW

**Comment by gpetiot** [comment]
I think it is ready to be reviewed now.

No regressions with `break-separators=after,dock-collection-brackets=false` and `break-separators=before,dock-collection-brackets=false`.

`break-separators=after,dock-collection-brackets=true` and `break-separators=before,dock-collection-brackets=true` look better, except in presence of tuples:
```ocaml
 and do_not_precompile args cls def k =
   ( {
-      me = Pm { cases = cls; args; default = def };
-      matrix = as_matrix cls;
-      top_default = def;
-    },
+    me = Pm { cases = cls; args; default = def };
+    matrix = as_matrix cls;
+    top_default = def;
+  },
     k )
```
but there is another issue for that: #1009, I will have a look at it but if it requires a lot of changes we might as well do it in another PR.

File: test/passing/comments_args.ml [hunk 1]
```diff
@@ -12,6 +12,10 @@ let emit_wrapper_function =
     (* is_memoize_impl *)
     Rx.NonRx false
 
+test 5 eq_int64
```
Comment (Julow) [inline test/passing/comments_args.ml]: Did you mean to add a `;;` before this line ?

**Comment by gpetiot** [comment]
Rewriting most functions to return a pair `Fmt.t * Fmt.t` is more difficult to maintain and to get the expected output, maybe Josh is right and we should wait until we ave an intermediate ast representation for this? @Julow

**Comment by Julow** [comment]
Yes that is becoming a mess. What we need is not an intermediate ast but a box representation that can either be manipulated or that handle docking of boxes cleanly.

**Comment by gpetiot** [comment]
Alright, I'm closing this one in the meantime, we should restart from scratch once we have better boxes.

**Comment by jberdine** [comment]
FTR, the 'intermediate ast' suggested in https://github.com/ocaml-ppx/ocamlformat/issues/406 isn't an ast, but a tree structure representing the (eventual) nesting of boxes. With such a structure, I think that nesting could be manipulated, and docking handled uniformly.

**Pull request closed without merging**